clang-tools-extra/clangd/ParsedAST.cpp

```diff
@@ -702,6 +704,10 @@ ParsedAST::build(llvm::StringRef Filename, const ParseInputs &Inputs,
     log("Execute() failed when building AST for {0}: {1}", MainInput.getFile(),
         toString(std::move(Err)));
 
+  // Disable the macro collector for the remainder of this function, e.g.
+  // clang-tidy checkers.
+  MacroCollectorPtr->doneParse();
```

Comment:
The first approach I tried was to remove `CollectMainFileMacros` from the preprocessor callbacks at this point, however the current `Preprocessor` API does not make that easy to do (adding a new callback creates a `PPChainedCallbacks`, and there are further internal calls made to `addPPCallbacks()` that chain other ones on top of ours). So, I opted to "neutralize" our callback rather than try to remove it.

Reply:
thanks it's unfortunate but i guess makes sense.

in theory, we can have similar issues with all the others PPCallbcks that we installed. Moreover if others were to use clang-tidy as a library, they'll run into similar pitfalls. As the flow is:
- Inject PPCallbacks
- Build AST
- Pass AST to clang-ast-consumer (in this case clang-tidy)
- Invoke EndOfMainFile

So there might be some value in injecting an extra callback between `Build AST` and `Pass AST to consumer`. We could properly reset our PPCallbacks to recognize leaving main file for such situations. It's unfortunate that both `FileChanged` and `LexedFileChanged` are designed to operate with a contract that hints "new file/location" will be valid. It makes such a semantic possibly breaking. Leaving that idea here in case you want to follow up on that (I'd be happy to review), but I can see that it's much more involved, possibly without anything breaking (and if it does we can always ask people to turn that check off until we fix the issue).

Reply:
Just to make sure I understand correctly, the idea is:

 * add a new method to `PPCallbacks`, e.g. `BuildASTDone()` or such
 * add a call to this new method in [place in libTooling, e.g. FrontendAction or such]
 * override `CollectMainFileMacros::BuildASTDone()` to do what `doneParse()` in the current patch does
 
?

Reply:
> add a new method to PPCallbacks, e.g. BuildASTDone() or such

I'd first see if we can extend existing `(Lexed)FileChanged` callbacks to fit this use case without breaking any users.

but if that doesn't work, yes, a new callback would be needed.